### DISCUSSION TITLE
Taking Effects Seriously

### DISCUSSION THREAD

**Discussion by zenna** [riadas/Autumn.jl#6] [General]
Currently in Autumn you can do the following:

```elm
x = 7
on event
  x => (prev x) * 2
```

There are a few shortcomings.  Suppose we want to encapsulate that behaviour

```elm
x = 7
double a = a => (prev a) * 2
on event
  double x
```

or, more usefully

```elm
double x = (prev x) * 2
x = 7
y = 7
xs = [x]
ys = []
on event
  move x xs ys -- move x from xs to ys
```

We cannot!  Why?
1. The `x => y` as the second argument to `on` is special; our examples are syntactically invalid
2. Inside `double` we can't access the previous value of `x`, that's only allowed at global scope
3. From the perspective of double and according to standard functional programming semantics, `x`is indistunguishable from `y` 

## Possible Solution
First, let's consider a simple Autumn program

```elm
x = init 1 next (prev x) + 1
on x > 10
  x => 1
```

We can think of this program as defining a sequence of `S = (s_1, s_2, ...)` where the state space is just `x`.  Actually, we can first consider a simpler sequence `Ŝ` of the program without the on clauses, which is essentially:

`Ŝ = [(x = 1,), (x = 2,), (x = 3,), ...]`

The on clause essentially modifies the sequence:

`S = [if pred(s) then f(s) else s for s in S]`

From the perspective we can consider on clauses as effectively defining `pred: S -> Bool` and `f: S -> S` which the compiler/interpreter uses to construct `S` from `Ŝ`

We could then rewrite the example to be:

```elm
x = init 1 next (prev x) + 1
on \s -> s.x > 10
  \s -> (x = s.x + 1,)
```

Of course this is verbose and cumbersome, 

## Syntax for effects

TLDR:
1. For any structure `x`, `x |> .a.b.c[1].d => d_` should return `x` with its `a.b.c[1].d`th element updated to `d_`
2. `.a.b.c[1].d => d_` should return a function which when applied to `x` is  `x` with its `a.b.c[1].d`th element updated to `d_`

```elm
on x > 10
  .x => 1
```

`.x => 1` is the function   `\s -> (x = s.x + 1,)`

```elm
move x xs ys = 
  let
    a = xs => remove x xs -- \s -> (xs = s.xs
    b = ys => add x xs
  in
    b ∘ a

x = 7
y = 7
xs = [x]
ys = []
on event
  move x .xs .ys
```

If you have a structure

```elm
struct Point
  a : A
  b : B
```

Then we can conceive of some autogenerated functions

```elm
geta : Point -> A
getb : Point -> B 
```

But also some update functions

```elm
upa : Point -> A -> Point
upa p a = Point a p.b
upb : Point -> B -> Point
upb p b = Point p.a b
```

Then `p.a => a_` is just sugar for `upa p a_` or maybe `upa a_`

If

```elm
struct A
  c : C
  d : D
```

Then `p.a.c => c_` is just sugar for `upa (upc c_)` [NOT QUITE RIGHT]
